### PR TITLE
Windows Installer: bump wix to v5

### PR DESCRIPTION
**Pull request by lsm5**
The chocolatey tool that was fetching us wix v3 can no longer be used to
fetch wix v4+ so we had to switch to dotnet to fetch the latest wix.
    
This commit builds the installer with wix v5.
wix v5 is installed via the `dotnet` tool in the windows image itself
at https://github.com/containers/automation_images/pull/354.
    
Going forward, the `dotnet` tool will also be used to build the installer.
    
In the process, the wix v3 files were converted to wix v4+ using `wix
convert` followed by manual modifications along with switch to wixproj
builds with dotnet.

The GitHub Action to upload windows installer now builds the installer
using winmake.ps1.
    
Contributions from Mario Loriedo:
- bundle setup update to wix5
- updates to build and release process scripts
Ref: https://github.com/lsm5/podman/pull/3

- small fixes to windows installer theme
Ref: https://github.com/lsm5/podman/pull/4

- Better win-installer sidebar logo
Ref: https://github.com/lsm5/podman/pull/5

    
Resolves: RUN-2055

#### Does this PR introduce a user-facing change?

```release-note
Windows installer is now built with WiX v5.
```

### CODE REVIEW

**Comment by ashley-cui** [comment]
The wix version probably also needs to be updated here, as it looks like we're hardcoding WiX 3.14

https://github.com/containers/podman/blob/60e0ef3e0d511d0250f6a988ff44deb0dfb73b6a/contrib/win-installer/build.ps1#L102

**Comment by lsm5** [comment]
> The wix version probably also needs to be updated here, as it looks like we're hardcoding WiX 3.14
> 
> https://github.com/containers/podman/blob/60e0ef3e0d511d0250f6a988ff44deb0dfb73b6a/contrib/win-installer/build.ps1#L102

@ashley-cui ugh, do you know if we can wildcard / skip it ? I'm hoping this way we can just use the version that dotnet installs and be done with it.

**Comment by ashley-cui** [comment]
@lsm5 I haven't tinkered with it enough to know unfortunately :(

**Comment by lsm5** [comment]
@cevich any idea why the Windows Cross task is failing? I updated the image SFX with what was generated in https://github.com/containers/automation_images/pull/354#issuecomment-2099070879

File: contrib/win-installer/build.ps1
Comment (cevich) [inline contrib/win-installer/build.ps1:130]: Drive-by comment: Are you sure this works in powershell?

In *nix the `*` must be on the outside of the quotes otherwise it's taken literally.

Comment (lsm5) [reply contrib/win-installer/build.ps1:135]: ack thanks I'll look into this. I've had trouble finding `dotnet` command from the windows image generated in automation_images so I'm working on that for now.

**Comment by lsm5** [comment]
igonre packit failures

**Comment by l0rd** [comment]
No, I haven't looked at the migration to WiX 4/5 but I would expect changes to the wxs/xml files (the schemas have changed). And my PR that is currently under review has some changes to those files so, if possible, use my branch as the base for your work.

**Comment by lsm5** [comment]
@Luap99 do we need the machine label on this PR? It updates the windows installer and the windows build and release process. The label was added by github action almost 2 months ago, but I don't know if it's still relevant.

**Comment by Luap99** [comment]
> @Luap99 do we need the machine label on this PR? It updates the windows installer and the windows build and release process. The label was added by github action almost 2 months ago, but I don't know if it's still relevant.

The label is set based on what file is changed by github action, I guess the action does't remove it again when someone repushes without machine changes?!
It has nothing do with what CI runs so feel free to drop that label.

As you touch .cirrus.yml all test have to run no matter what.

**Comment by lsm5** [comment]
@ashley-cui @l0rd @n1hility @baude PTAL.
cc @containers/podman-maintainers

**Comment by lsm5** [comment]
/hold

setting hold so that we merge https://github.com/containers/automation_images/pull/354 before merging this.

**Comment by lsm5** [comment]
> LGTM I have been able to test different scenarios (with and without wsl/hyperv, setting/unsetting WSL/Hyper-V checkboxes) and it worked as expected.
> 
> [I have pushed another commit](https://github.com/lsm5/podman/pull/4) with some minor improvements (mainly to the setup.exe GUI). I would like to include it in this PR but it can also be submitted as a separate one.

I'll include it here. Thanks!

**Comment by lsm5** [comment]
setting this back to draft to test out a change to windows image with simplified dotnet install.

**Comment by lsm5** [comment]
@l0rd windows jobs are passing with the latest version of https://github.com/containers/automation_images/pull/354 . PTAL again.

**Comment by lsm5** [comment]
looks like the `machine-mac podman darwin rootless host sqlite` test is failing on other PRs too. @Luap99 ok to ignore it or should we wait for a fix?

**Comment by Luap99** [comment]
> looks like the `machine-mac podman darwin rootless host sqlite` test is failing on other PRs too. @Luap99 ok to ignore it or should we wait for a fix?

Unfortunately we have to re-run until it passes. And yes it took 12 tries on my PR :(
I try to debug/fix this currently but it is more complex than thought: https://github.com/containers/podman/pull/23162

**Review by l0rd — APPROVED** [review]
LGTM

**Comment by lsm5** [comment]
@cevich do we need the optional release test on this PR?

**Comment by lsm5** [comment]
i'll need another run of this once the latest build of automation_images is done.

**Comment by cevich** [comment]
> do we need the optional release test on this PR?

No, definitely not, it's guaranteed to fail :wink:

**Comment by lsm5** [comment]
Guess this is good to go. Needs a slash-lgtm unless any other comments.

@containers/podman-maintainers PTAL

File: .packit.yaml
Comment (Luap99) [inline .packit.yaml:78]: That should not be merged, should it?

File: contrib/win-installer/resources/podman-sidebar.png
Comment (Luap99) [inline contrib/win-installer/resources/podman-sidebar.png:1]: Why is this changed? And why does it have a black bar at the bottom of the new image. If the size needs to be changed I would expect it to have at least the consistent background color.

**Review by Luap99 — CHANGES_REQUESTED** [review]
Some things I noticed that do not look right. I cannot review any of the windows/installer changes though but I guess if it works it works.

File: contrib/win-installer/resources/podman-sidebar.png
Comment (Luap99) [reply contrib/win-installer/resources/podman-sidebar.png:1]: I mean if it looks correct in the installer then it is fine. But it just looks strange to have a black bar in the image and not fill up the rest with the same background color if we want the new size.

**Comment by edsantiago** [comment]
@lsm5 ping, please address Paul's question, rebase, and repush.

File: contrib/win-installer/resources/podman-sidebar.png
Comment (l0rd) [reply contrib/win-installer/resources/podman-sidebar.png:1]: @Luap99 the installer shows the whole image without any modification so I guess that's visible. But I cannot spot the difference unless I zoom in so I cannot really tell. I have pushed an updated version of the image [here](https://github.com/lsm5/podman/pull/5/files), please have a look.

**Comment by lsm5** [comment]
all done. good to go. PTAL @containers/podman-maintainers

**Comment by baude** [comment]
LGTM

**Review by cevich — APPROVED** [review]
Non-expert LGTM.  I really dislike all the hard-coded version numbers.  I briefly checked, but don't see any non-easy way to manage them with Renovate.  I'm assuming this is simply "the way" of this particular environment.  Thanks everyone for your hard work on this, it's been a long time coming!

**Comment by lsm5** [comment]
> Non-expert LGTM. I really dislike all the hard-coded version numbers. I briefly checked, but don't see any non-easy way to manage them with Renovate. I'm assuming this is simply "the way" of this particular environment. Thanks everyone for your hard work on this, it's been a long time coming!

I guess we could use wildcards for the Package references in `.wixproj` files and I think that will use the latest available versions, but I don't know if we'd end up with build job failures. We could do followup PRs to address that.

**Review by ashley-cui — COMMENTED** [review]
I was following the doc on how to build the installer, but ran into this issue. Where can I get this dependency?

```
.\winmake.ps1 installer                                      
 Building the windows installer                                                                                         
  .\build.ps1 5.2.0 dev "C:\Users\Ashley\go\src\github.com\containers\podman\contrib\win-installer\current"               
  Required dep "MingW CC" is not installed            
```

Also, does this affect any part of our release process? Or does the GitHub Action function as before?

Thanks for all the work, @lsm5

**Comment by lsm5** [comment]
> I was following the doc on how to build the installer, but ran into this issue. Where can I get this dependency?
> 
> ```
> .\winmake.ps1 installer                                      
>  Building the windows installer                                                                                         
>   .\build.ps1 5.2.0 dev "C:\Users\Ashley\go\src\github.com\containers\podman\contrib\win-installer\current"               
>   Required dep "MingW CC" is not installed            
> ```
> 
> Also, does this affect any part of our release process? Or does the GitHub Action function as before?

Actually good question about GHA. I should run a check for that before we merge. Thanks @ashley-cui

**Comment by lsm5** [comment]
> > I was following the doc on how to build the installer, but ran into this issue. Where can I get this dependency?
> > ```
> > .\winmake.ps1 installer                                      
> >  Building the windows installer                                                                                         
> >   .\build.ps1 5.2.0 dev "C:\Users\Ashley\go\src\github.com\containers\podman\contrib\win-installer\current"               
> >   Required dep "MingW CC" is not installed            
> > ```

@l0rd did you see this at all on your env? Wonder if it's a path issue. 

> > Also, does this affect any part of our release process? Or does the GitHub Action function as before?

Once we verify the action works, I've been thinking if it would be good to have the windows installer dryrun on PRs or if it would consume a lot of our actions quota?

**Comment by cevich** [comment]
> Once we verify the action works, I've been thinking if it would be good to have the windows installer dryrun on PRs or if it would consume a lot of our actions quota?

My strong preference would be to do as much as humanly possible in Cirrus-CI.  Only use GHA for short/quick things or items that absolutely must operate with secure secrets (like signing).  We're very stretched on the GHA 20-workers quota available for the entire containers-org.  The "fix" to that is extraordinarily complex and potentially hazardous even.

Also, doesn't testing the installer require running on bare-metal anyway?

**Comment by l0rd** [comment]
> @l0rd did you see this at all on your env? Wonder if it's a path issue.

Yes. @ashley-cui is right. But that's unrelated to this PR. That's a pre-exising problem and the CI isn't affected: [MingW is installed](https://github.com/containers/automation_images/blob/main/win_images/win_packaging.ps1#L12) in `automation_images`. The Windows installer tests ([1](https://cirrus-ci.com/task/5818848429998080) and [2](https://cirrus-ci.com/task/5255898476576768)) are passing.

This is a documentation problem but I was able to locally use the [Microsoft C compiler](https://learn.microsoft.com/en-us/cpp/build/building-on-the-command-line?view=msvc-170) instead of `gcc` to build `contrib/win-installer/podman-msihooks/check.c`. I am working on a separate PR to update `windows_build.md` and get rid of MingW once and for all 😄

**Comment by ashley-cui** [comment]
> > Once we verify the action works, I've been thinking if it would be good to have the windows installer dryrun on PRs or if it would consume a lot of our actions quota?
> 
> My strong preference would be to do as much as humanly possible in Cirrus-CI. Only use GHA for short/quick things or items that absolutely must operate with secure secrets (like signing). We're very stretched on the GHA 20-workers quota available for the entire containers-org. The "fix" to that is extraordinarily complex and potentially hazardous even.
> 
> Also, doesn't testing the installer require running on bare-metal anyway?

This makes sense, I'd agree with only running the GHA on releases, since that's necessary for signing the pkg.

**Comment by lsm5** [comment]
@ashley-cui @l0rd so are we ok to skip all GHA changes in this PR? I can remove my additional commit in that case

**Comment by lsm5** [comment]
> @ashley-cui @l0rd so are we ok to skip all GHA changes in this PR? I can remove my additional commit in that case

Talked to @l0rd privately just now. GHA doesn't have wix v5 installed by default so we definitely to account for that in this PR or else we'll only find out at release time.

I'll try to get the windows GHA to a point where we `can` run it on PRs if we want to, but have that part disabled by default.

**Comment by l0rd** [comment]
> I was following the doc on how to build the installer, but ran into this issue. Where can I get this dependency?
> 
> ```
> .\winmake.ps1 installer                                      
>  Building the windows installer                                                                                         
>   .\build.ps1 5.2.0 dev "C:\Users\Ashley\go\src\github.com\containers\podman\contrib\win-installer\current"               
>   Required dep "MingW CC" is not installed            
> ```

I have added [this commit (based on this PR)](https://github.com/l0rd/podman/commit/08e08a07ac0f1f6561554549701d015d14f8061c) to address the MingW CC problem. I will submit a dedicated PR when this one get merged.

**Comment by ashley-cui** [comment]
LGTM, thanks!

**Review by l0rd — APPROVED** [review]
LGTM 👍 thank you @lsm5

File: .github/workflows/upload-win-installer.yml
Comment (l0rd) [inline .github/workflows/upload-win-installer.yml:113]: `podman` is a `podman-remote` alias. It's built twice the same binary here.
```suggestion
```

Comment (lsm5) [reply .github/workflows/upload-win-installer.yml]: Done, thanks!

**Comment by cevich** [comment]
Paul's https://github.com/containers/podman/pull/23162 will address the flaking PM Mac test.  Or we can just re-run it 5 or 10 times :weary:

**Comment by lsm5** [comment]
> Paul's #23162 will address the flaking PM Mac test. Or we can just re-run it 5 or 10 times 😩

i'll rebase on that one. Thanks!

**Comment by cevich** [comment]
K.  It should merge soon.

**Comment by cevich** [comment]
Thanks for all your persistence on this @lsm5 :smile:

**Pull request merged**
